File: extensions/h5/properties/FAPL.cc
// -*- C++ -*-
// -*- coding: utf-8 -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2026 all rights reserved


// externals
#include "external.h"
// namespace setup
#include "forward.h"


// file access property lists
void
pyre::h5::py::properties::fapl(py::module & m)
{
    // add bindings for hdf5 file access property lists
    auto cls = py::class_<FAPL, PropList>(
        // in scope
        m,
        // class name
        "fapl",
        // docstring
        "a file access property list");

    // static properties
    cls.def_property_readonly_static(
        // the name
        "default",
        // the implementation
        [](const py::object &) -> const FAPL & {
            // easy enough
            return FAPL::theDefault();
        },
        // we hand back a reference to a shared, library-owned object
        py::return_value_policy::reference,
        // docstring
        "the default file access property list");

    // constructor
    cls.def(
        // the implementation
        py::init(),
        // the docstring
        "create a file access property list");

    // interface
    // the metadata block size
    cls.def_property(
        // the name
        "metadataBlockSize",
        // the getter
        py::overload_cast<>(&FAPL::metadataBlockSize, py::const_),
        // the setter
        py::overload_cast<hsize_t>(&FAPL::metadataBlockSize),
        // the docstring
        "the size of the blocks my metadata is allocated in");

    // the page buffer characteristics
    cls.def_property(
        // the name
        "pageBufferSize",
        // the getter
        py::overload_cast<>(&FAPL::pageBufferSize, py::const_),
        // the setter
        py::overload_cast<const PageBuffer &>(&FAPL::pageBufferSize),
        // the docstring
        "my page buffer, as {(bytes, metadata percent, raw data percent)}");

    // the alignment of objects in the file
    cls.def_property(
        // the name
        "alignment",
        // the getter
        py::overload_cast<>(&FAPL::alignment, py::const_),
        // the setter
        py::overload_cast<const Alignment &>(&FAPL::alignment),
        // the docstring
        "where objects start in my file, as {(threshold, alignment)} bytes");

    // the sieve buffer
    cls.def_property(
        // the name
        "sieveBufferSize",
        // the getter
        py::overload_cast<>(&FAPL::sieveBufferSize, py::const_),
        // the setter
        py::overload_cast<std::size_t>(&FAPL::sieveBufferSize),
        // the docstring
        "the size of the buffer that gathers small writes to contiguous datasets");

    // the file close degree
    cls.def_property(
        // the name
        "closeDegree",
        // the getter
        py::overload_cast<>(&FAPL::closeDegree, py::const_),
        // the setter
        py::overload_cast<H5F_close_degree_t>(&FAPL::closeDegree),
        // the docstring
        "what becomes of my file when its handle closes with objects still open");

    // the default caches
    cls.def_property(
        // the name
        "cache",
        // the getter
        py::overload_cast<>(&FAPL::cache, py::const_),
        // the setter
        py::overload_cast<const Cache &>(&FAPL::cache),
        // the docstring
        "my default caches, as {(metadata elements, chunk slots, chunk bytes, preemption)}");

    // the file format version bounds
    cls.def_property(
        // the name
        "libverBounds",
        // the getter
        py::overload_cast<>(&FAPL::libverBounds, py::const_),
        // the setter
        py::overload_cast<const VersionBounds &>(&FAPL::libverBounds),
        // the docstring
        "the file format versions i may use, as {(low, high)}; asking for a recent one "
        "buys newer features and narrows who can read the result");

#if defined(H5_HAVE_ROS3_VFD)
    // populate the property list with ros3 parameters
    cls.def(
        // the name
        "ros3",
        // the implementation
        &FAPL::ros3,
        // the signature
        "authenticate"_a = true, "region"_a = "", "id"_a = "", "key"_a = "", "token"_a = "",
        // we hand back a reference to the list we just configured
        py::return_value_policy::reference,
        // the docstring
        "populate the property list with ros3 parameters");
#endif

    // all done
    return;
}


// end of file

File: tests/h5.ext/fapl.py
#!/usr/bin/env python3
# -*- Python -*-
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2026 all rights reserved


def test():
    """
    Exercise the file access property list: the settings that govern how a file is reached,
    which apply on every open rather than once at creation
    """
    # get the bindings
    from pyre.extensions import libh5

    # for a scratch path and its cleanup
    import os

    # a scratch data product
    uri = "h5_ext_fapl.h5"
    # make sure a stale one is not lying around
    if os.path.exists(uri):
        os.remove(uri)

    # a fresh file access property list
    fapl = libh5.properties.fapl()

    # objects are not aligned, by default: any position will do
    assert fapl.alignment == (1, 1)
    # a product headed for a parallel filesystem lands its larger objects on boundaries
    fapl.alignment = (4096, 4096)
    assert fapl.alignment == (4096, 4096)

    # the sieve buffer gathers small writes to contiguous datasets
    assert fapl.sieveBufferSize > 0
    # and it is ours to size
    fapl.sieveBufferSize = 128 * 1024
    assert fapl.sieveBufferSize == 128 * 1024

    # the caches, as {(metadata elements, chunk slots, chunk bytes, preemption)}; these are
    # the defaults a dataset access property list overrides one dataset at a time
    elements, slots, bytes, w0 = fapl.cache
    # a workflow that sweeps large chunks wants room for more of them
    fapl.cache = (elements, slots, 64 * 1024 * 1024, w0)
    assert fapl.cache == (elements, slots, 64 * 1024 * 1024, w0)

    # what becomes of a file closed while objects in it are still open
    assert fapl.closeDegree == libh5.CloseDegree.default
    # ask for the handle to survive until the last object is done with it
    fapl.closeDegree = libh5.CloseDegree.strong
    assert fapl.closeDegree == libh5.CloseDegree.strong

    # the file format versions the library may use; {latest} is not a fixed point, it means
    # whatever the library we were built against considers newest, so reading the bound
    # back reports the concrete version it resolved to rather than the request
    fapl.libverBounds = (libh5.LibVersion.earliest, libh5.LibVersion.latest)
    low, high = fapl.libverBounds
    # the low bound is what we asked for
    assert low == libh5.LibVersion.earliest
    # and the high bound is at least as new as the oldest named format
    assert high >= libh5.LibVersion.v18

    # pinning the format keeps older libraries able to read what we write
    fapl.libverBounds = (libh5.LibVersion.earliest, libh5.LibVersion.v18)
    assert fapl.libverBounds == (libh5.LibVersion.earliest, libh5.LibVersion.v18)

    # make a file with it
    f = libh5.File(uri=uri, mode="w", fapl=fapl)
    # and put something in it, so the file is worth reading back
    f.create(path="group")

    # all done
    return


# main
if __name__ == "__main__":
    # drive
    test()


# end of file
